Web Inspector: Network panel: Disable context menu items for incomplete resources

rcaliman-apple opened the pull request:
#### b23b8d5f212a59acaa618428d9e613f5757451fb
<pre>
Web Inspector: Network panel: Disable context menu items for incomplete resources
<a href="https://bugs.webkit.org/show_bug.cgi?id=242035">https://bugs.webkit.org/show_bug.cgi?id=242035</a>

Reviewed by Devin Rousso.

Information collected about resources loaded before Web Inspector was opened is missing most metadata,
like resource timing data, HTTP method, request headers, etc.

* Source/WebInspectorUI/UserInterface/Models/Resource.js:
(WI.Resource.prototype.get hasMetadata):

Add a new member to quickly check if the resource has metadata.

* Source/WebInspectorUI/UserInterface/Views/ContextMenuUtilities.js:

In absence of metadata, context menu items to copy as cURL, copy as fetch, copy HTTP request
and copy HTTP response should not be shown to avoid getting useless output.

Canonical link: <a href="https://commits.webkit.org/251987@main">https://commits.webkit.org/251987@main</a>
</pre>

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ContextMenuUtilities.js:
```diff
@@ -178,24 +178,28 @@ WI.appendContextMenuItemsForSourceCode = function(contextMenu, sourceCodeOrLocat
         if (sourceCode.urlComponents.scheme !== "data") {
+
+            // Information collected about resources loaded before Web Inspector was opened is missing most metadata, like resource timing data, HTTP method, request headers, etc.
+            let isDisabled = !sourceCode.requestSentDate;
```
Let's create a `WI.Resource.prototype.hasMetadata` (or something with a similar/better name) and use that instead (and maybe for HAR export too).  I don't think we should propagate `requestSentDate` (and the comment) every time we need to ask "does this `WI.Resource` have the full set of details needed to do the thing).

rcaliman-apple replied:
Will do.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/NetworkTableContentView.js:
```diff
@@ -1396,7 +1396,7 @@ WI.NetworkTableContentView = class NetworkTableContentView extends WI.ContentVie
             return false;
 
-        if (!mainResource.requestSentDate)
+        if (!mainResource.hasMetadata)
```
So I looked at `WI.HARBuilder` and I think technically the only thing that's actually required for a `WI.Resource` is the `requestSentDate`.  All of the other information (e.g. `timingData`, `remoteAddress`, etc.) is optional.

As such, I'd probably undo this change and leave it as it was before (or maybe instead create a `WI.Resource.prototype.hasHARData` or something).  My apologies for suggesting you change it.

rcaliman-apple replied:
Done. Removed

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/NetworkTableContentView.js:
```diff
@@ -2394,7 +2394,7 @@ WI.NetworkTableContentView = class NetworkTableContentView extends WI.ContentVie
             if (!resource.finished)
                 return false;
-            if (!resource.requestSentDate)
+            if (!resource.hasMetadata)
```
ditto (:1399)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ContextMenuUtilities.js:
```diff
@@ -178,24 +178,27 @@ WI.appendContextMenuItemsForSourceCode = function(contextMenu, sourceCodeOrLocat
     if (sourceCode instanceof WI.Resource && !sourceCode.localResourceOverride) {
         if (sourceCode.urlComponents.scheme !== "data") {
+
+            let isDisabled = !sourceCode.hasMetadata;
```
i kinda feel like we shouldn't even show these contextmenu items if we don't have the data necessary to perform the action (e.g. that's normally what we do when the action relies on protocol changes), just like how we don't even show them for `data:` resources 🤔

rcaliman-apple replied:
Ok. Not showing context menu items for resources without metadata.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/Resource.js:
```diff
@@ -435,6 +435,14 @@ WI.Resource = class Resource extends WI.SourceCode
+    {
+        // Information collected about resources loaded before Web Inspector was opened is missing most metadata,
+        // like resource timing data, HTTP method, request headers, etc.
+        // Missing a requestSentDate is an indicator that other important metadata is also missing.
```
these sentences are kinda awkward to read.  i'd reword them something along the lines of
```
// Some metadata is only collected when Web Inspector is open (e.g. resource timing data, HTTP method, request headers, etc.).
// Use `requestSentDate` as a general signal since it is always included when metadata is collected.
```

also, i'm not sure this was provided before iOS 11 (see `WI.NetworkObserver.prototype.requestWillBeSent`) so we might need to find a different general signal 😕  maybe `requestIdentifier`?

rcaliman-apple replied:
Updated. `requestIdentifier` works as a signal

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/Resource.js:
```diff
@@ -435,6 +435,13 @@ WI.Resource = class Resource extends WI.SourceCode
+    {
+        // Some metadata is only collected when Web Inspector is open (e.g. resource timing data, HTTP method, request headers, etc.).
+        // Use `requestIdentifier` as a general signal since it is always included when metadata is collected.
+        return !!this.requestIdentifier;
```
please use `this._requestIdentifier` since this is inside the same class that declares it (i.e. it's not a layering violation)

rcaliman-apple committed:
Web Inspector: Network panel: Disable context menu items for incomplete resources
https://bugs.webkit.org/show_bug.cgi?id=242035

Reviewed by Devin Rousso.

Information collected about resources loaded before Web Inspector was opened is missing most metadata,
like resource timing data, HTTP method, request headers, etc.

* Source/WebInspectorUI/UserInterface/Models/Resource.js:
(WI.Resource.prototype.get hasMetadata):

Add a new member to quickly check if the resource has metadata.

* Source/WebInspectorUI/UserInterface/Views/ContextMenuUtilities.js:

In absence of metadata, context menu items to copy as cURL, copy as fetch, copy HTTP request
and copy HTTP response should not be shown to avoid getting useless output.

Canonical link: https://commits.webkit.org/251987@main
